# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/JSAtomUtils.cpp
# Commit: ae6126131b6d
# Full Hash: ae6126131b6dfa18981f25c6986200c14e63e5f2
# Author: Doug Thayer <dothayer@mozilla.com>
# Date: 2024-04-11 04:26:26
# Regressor Bug: 1881995
# File Overlap Count: 1
# Description:
#   Bug 1881995 - Implement ForwardedAtoms and create them during atomization r=iain,sfink
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D202690
# ==============================================================================

diff -r 4e5967329fdd -r ae6126131b6d js/src/vm/JSAtomUtils.cpp
--- a/js/src/vm/JSAtomUtils.cpp	Wed Apr 10 15:47:15 2024 +0000
+++ b/js/src/vm/JSAtomUtils.cpp	Wed Apr 10 16:32:23 2024 +0000
@@ -666,6 +666,10 @@
     return &str->asAtom();
   }
 
+  if (str->isAtomRef()) {
+    return str->atom();
+  }
+
   if (JSAtom* atom = cx->caches().stringToAtomCache.lookup(str)) {
     return atom;
   }
@@ -691,6 +695,7 @@
         // not done in lookup() itself, because #including JSContext.h there
         // causes some non-trivial #include ordering issues.
         cx->markAtom(atom);
+        str->tryReplaceWithAtomRef(atom);
         return atom;
       }
     }
@@ -723,7 +728,9 @@
     return nullptr;
   }
 
-  cx->caches().stringToAtomCache.maybePut(str, atom, key);
+  if (!str->tryReplaceWithAtomRef(atom)) {
+    cx->caches().stringToAtomCache.maybePut(str, atom, key);
+  }
 
   return atom;
 }